File: runtime/src/test/java/com/siimkinks/sqlitemagic/ComplexColumnTest.java
package com.siimkinks.sqlitemagic;

import androidx.annotation.NonNull;

import org.junit.Test;

import java.util.Arrays;

import static com.siimkinks.sqlitemagic.ComplexColumnTestFixtures.ARTICLE;
import static com.siimkinks.sqlitemagic.ComplexColumnTestFixtures.Article;
import static com.siimkinks.sqlitemagic.SqlTestAssertions.assertExpression;
import static com.siimkinks.sqlitemagic.Utils.STRING_PARSER;

public final class ComplexColumnTest {
  @Test
  public void textAffinityUsesDeclaredIdTypeAndTransformer() {
    final ComplexColumn<AccountId, AccountId, CharSequence, Article, NotNullable> column =
        new AccountTextColumn();

    assertExpression(column.is(new AccountId("first")), "article.account=?", "db:first");
    assertExpression(
        column.in(Arrays.asList(new AccountId("first"), new AccountId("second"))),
        "article.account IN (?,?)",
        "db:first",
        "db:second"
    );
  }

  @Test
  public void genericComplexColumnRetainsNullPredicates() {
    final ComplexColumn<AccountId, AccountId, CharSequence, Article, Nullable> column =
        new NullableAccountTextColumn();

    assertExpression(column.isNull(), "article.account IS NULL");
    assertExpression(column.isNotNull(), "article.account IS NOT NULL");
  }

  private record AccountId(String value) {
  }

  private static class AccountTextColumn extends ComplexColumn<AccountId, AccountId, CharSequence, Article, NotNullable> {
    AccountTextColumn() {
      super(ARTICLE, "account", false, STRING_PARSER, false, null);
    }

    @NonNull
    @Override
    String toSqlArg(AccountId value) {
      return "db:" + value.value;
    }
  }

  private static final class NullableAccountTextColumn extends ComplexColumn<AccountId, AccountId, CharSequence, Article, Nullable> {
    NullableAccountTextColumn() {
      super(ARTICLE, "account", false, STRING_PARSER, true, null);
    }

    @NonNull
    @Override
    String toSqlArg(AccountId value) {
      return "db:" + value.value;
    }
  }
}

File: runtime/src/test/kotlin/com/siimkinks/sqlitemagic/ComplexColumnTest.kt
package com.siimkinks.sqlitemagic

import com.siimkinks.sqlitemagic.ComplexColumnTestFixtures.ARTICLE
import com.siimkinks.sqlitemagic.ComplexColumnTestFixtures.Article
import com.siimkinks.sqlitemagic.SqlTestAssertions.assertExpression
import org.junit.Test

class ComplexColumnTest {
  @Test
  fun `text affinity uses declared ID type and transformer`() {
    val column = AccountTextColumn()

    assertExpression(
      expression = column.`is`(AccountId("first")),
      expectedSql = "article.account=?",
      "db:first"
    )
    assertExpression(
      expression = column.`in`(listOf(AccountId("first"), AccountId("second"))),
      expectedSql = "article.account IN (?,?)",
      "db:first",
      "db:second"
    )
  }

  @Test
  fun `generic complex column retains null predicates`() {
    val column = NullableAccountTextColumn()

    assertExpression(
      expression = column.isNull(),
      expectedSql = "article.account IS NULL"
    )
    assertExpression(
      expression = column.isNotNull(),
      expectedSql = "article.account IS NOT NULL"
    )
  }

  private data class AccountId(val value: String)

  private class AccountTextColumn : ComplexColumn<AccountId, AccountId, CharSequence, Article, NotNullable>(
    table = ARTICLE,
    name = "account",
    valueParser = Utils.STRING_PARSER,
    nullable = false,
    alias = null
  ) {
    override fun toSqlArg(value: AccountId) = "db:${value.value}"
  }

  private class NullableAccountTextColumn :
    ComplexColumn<AccountId, AccountId, CharSequence, Article, Nullable>(
      table = ARTICLE,
      name = "account",
      valueParser = Utils.STRING_PARSER,
      nullable = true,
      alias = null
    ) {
    override fun toSqlArg(value: AccountId) = "db:${value.value}"
  }
}
